src/puya/puya-binary.ts

```diff
@@ -0,0 +1,199 @@
+  const archiveUrl = `https://github.com/${Constants.puyaGithubRepo}/releases/download/v${version.formatted}/${archiveFileName}`
+  const checksumUrl = `https://github.com/${Constants.puyaGithubRepo}/releases/download/v${version.formatted}/${checksumFileName}`
+
+  await downloadFile(archiveUrl, archiveFilePath)
+  await downloadFile(checksumUrl, checksumFilePath)
+
+  await verifyChecksum(archiveFilePath, archiveFileName, checksumFilePath)
+
+  await tar.extract({
+    file: archiveFilePath,
+    cwd: puyaStorageDir,
+  })
+
+  // Check if extraction was successful and binary exists
+  if (!fs.existsSync(extractedBinaryPath)) {
+    throw new InternalError(`Binary file ${binaryFileName} not found in the extracted archive`)
+  }
+
+  // Make binary executable on non-Windows platforms
+  if (os !== 'windows') {
+    fs.chmodSync(extractedBinaryPath, 0o755)
+  }
+
+  logger.debug(undefined, `Successfully downloaded and extracted Puya binary to ${extractedBinaryPath}`)
+  return extractedBinaryPath
+}
+
+/**
+ * Downloads a file from a URL to a specified path
+ * @param url The URL to download from
+ * @param destination The local path to save the file to
+ * @returns Promise that resolves when the download is complete
+ */
+async function downloadFile(url: string, destination: string): Promise<void> {
+  const response = await fetch(url)
+  if (!response.ok) {
+    throw new Error(`Failed to download file. Status Code: ${response.status}`)
+  }
+  const buffer = await response.arrayBuffer()
+  return await new Promise<void>((resolve, reject) => {
+    const fileStream = fs.createWriteStream(destination)
+    const nodeBuffer = Buffer.from(buffer)
+
+    fileStream.write(nodeBuffer)
+    fileStream.end()
+
+    fileStream.on('finish', () => {
+      resolve()
+    })
+
+    fileStream.on('error', (err) => {
+      fs.unlinkSync(destination)
+      reject(err)
+    })
+  })
+}
+
+/**
+ * Verifies the checksum of a downloaded file
+ * @param filePath Path to the file to verify
+ * @param checksumFilePath Path to the checksum file
+ * @returns Promise that resolves to true if the checksum matches, false otherwise
+ */
+async function verifyChecksum(filePath: string, fileName: string, checksumFilePath: string): Promise<void> {
+  return new Promise((resolve, reject) => {
+    // Read the checksum file
+    const expectedChecksum = fs.readFileSync(checksumFilePath, 'utf8').trim().toLowerCase()
+
+    // Calculate the SHA256 checksum of the downloaded file
+    const hash = crypto.createHash('sha256')
+    const stream = fs.createReadStream(filePath)
+
+    stream.on('data', (data) => {
+      hash.update(data)
+    })
+
+    stream.on('end', () => {
+      const calculatedChecksum = hash.digest('hex').toLowerCase()
+
+      if (`${calculatedChecksum}  ${fileName}` !== expectedChecksum) {
+        reject(new InternalError(`Checksum verification failed. Expected checksum: ${expectedChecksum} but got: ${calculatedChecksum}`))
+      }
+      resolve()
```

Comment:
```suggestion
      if (`${calculatedChecksum}  ${fileName}` === expectedChecksum) {
        resolve()
      }
      reject(new InternalError(`Checksum verification failed. Expected checksum: ${expectedChecksum} but got: ${calculatedChecksum}`))
```

This condition should probably be inverted, so that it fails by default.